<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html lang="en-us" xmlns="http://www.w3.org/1999/xhtml" >
<head>
	<meta charset="UTF-8" name="viewport" content="width=device-width, intial-scale=1">
	<title>{{ "%s %s %s"%(calculation_results.gene,calculation_results.unp,calculation_results.variant) }}</title>
	<!-- <link rel="stylesheet" href="https://www.w3schools.com/w3css/4/w3.css"> -->
 <link rel="stylesheet" href="https://www.w3schools.com/w3css/4/w3.css">
	<!-- jQuery -->
	<script type="text/javascript" src="../html/javascript/jquery-3.5.1/jquery-3.5.1.min.js"></script>

	<!-- Demo stuff from mottie.github.io -->
	<link rel="stylesheet" href="../html/javascript/tablesorter_docs/css/jq.css" type="text/css" media="print, projection, screen" />
	<link rel="stylesheet" href="../html/javascript/tablesorter_docs/css/prettify.css" type="text/css" media="print, projection, screen" />
    <script src="../html/javascript/tablesorter_docs/js/prettify.js"></script>
	<script src="../html/javascript/tablesorter_docs/js/docs.js"></script>

	<!-- Tablesorter required -->
	<link rel="stylesheet" href="../html/javascript/tablesorter/css/theme.blue.css">
	<script type="text/javascript" src="../html/javascript/tablesorter/js/jquery.tablesorter.js"></script>

	<!-- Tablesorter: optional -->
	<!-- <script type="text/javascript" src="../html/javascript/tablesorter/js/extras/jquery.tablesorter.pager.min.js"></script> -->


      <link rel="stylesheet" href="../html/css/typography.css" />
		<style media="print" type="text/css">
		h4{text-align:center;margin:0;padding:0.1in;}
		h5{text-align:center;margin:0;padding:0.1in;}
		.container{width:8in;float:left;}
		.box{width:2in;float:left;height:auto;}
		.big-box{width:3in;float:left;height:auto;}
		</style>

<script type="text/javascript">
	// In current javascript, syntax below does what we need
   $(function() {
                                $(".tablesorter").tablesorter({
                                        // sortReset: 'true',
                                        theme: 'blue'
                                });
        });

  function resizeIframe(obj) {
    obj.style.height = (40 + obj.contentWindow.document.body.scrollHeight) + 'px';
  }

   /* http://colorbrewer2.org/#type=qualitative&amp;scheme=Paired&amp;n=9
      Commented colors from chir.ag/projects/name-that-color
    */

</script>



    <!-- load the prototype library from google
    <script type="text/javascript" src="https://www.google.com/jsapi"></script>
    <script type="text/javascript">google.load("prototype", "1.7");</script>
    -->

    <!-- styles applicable to all browsers 
    <link rel='stylesheet' href='https://pfam.xfam.org/static/css/tab_layout.css' type='text/css' />
    <link rel='stylesheet' href='https://pfam.xfam.org/static/css/protein.css' type='text/css' />
	 -->

    <!-- in order to standardize the look of text, and in order to make it work at
         all in older browsers (and IE), we need the canvas text library. And, in turn,
         that library uses a font description in a separate javascript file -->
    <!--
      <script type="text/javascript" src="https://pfam.xfam.org/static/javascripts/canvas.text.js?dontUseMoz=true&amp;reimplement=true"></script>
    -->

    <!--
    <script type="text/javascript" src="https://pfam.xfam.org/static/javascripts/canvas.text.js?dontUseMoz=true&amp;reimplement=true"></script>
    <script type="text/javascript" src="https://pfam.xfam.org/static/javascripts/faces/optimer-bold-normal.js"></script>
    -->

    <!-- NGL molecular graphics viewer -->
    <script type="text/javascript" src="../html/javascript/ngl.js"></script>


    <!-- IE, being IE, needs extra help. In particular we need to add an extra library
         that implements <canvas> using the VML drawing layer in IE. And, so that 
         "console.log" calls don't screw everything up, we include the firebug-lite 
         script -->
    <!--[if IE]>
      <script type="text/javascript" src="https://pfam.xfam.org/static/javascripts/excanvas.js"></script>
      <script type="text/javascript" src="https://getfirebug.com/releases/lite/1.2/firebug-lite-compressed.js"></script>
    <![endif]-->

    <style type="text/css">
      #dg canvas { 
        margin: 2em;
      }
      #errors {
        border: 1px solid red;
        margin: 1em 0;
        padding: 0.4em;
      }

<style>
* {
    box-sizing: border-box;
}

/* Make grey header boxes for the sub-sections */
.header, .footer {
    background-color: grey;
    display: inline-block;
    /* text-align: center; */
    padding-top: 7px;
    padding-bottom: 0px;
    padding-left: 40px;
    padding-right: 40px;
    margin-top: 5px;
    
}

div.header * {
	color:white;
   vertical-align: middle;
}

body {
    margin: 0px;         /* Reset default margin on the body element */
    padding: 0px;
    height: 100%;
    overflow:hidden;
}
iframe {
    /* border: "1px solid black";         /* Reset default border */
    width: 100%;
    overflow: hidden;
    scrolling: "no";
}

.wrapper {
	position: relative;
	width: 100%; /* <-- just a default, can be anything. */
}

div.wrapper * {
	text-align: center;
}

img {
    max-width: 100%;
    max-height: 100%;
}

/* Positioning for the main box */
.main-box {
    width: 99%;
    margin: 1% auto;
    display: inline-block;
}


/* Create two equal columns that floats next to each other */
.column2 {
    /* float: left; */
    width: 50%;
    padding: 10px;
    /* height: 300px; /* Should be removed. Only for demonstration */
}
/* Create three equal columns that floats next to each other */
.column3 {
   float: left;
   width: 33%;
   padding: 8px;
	padding-bottom: 6px;
	/* display: inline-block; */
    /* height: 300px; /* Should be removed. Only for demonstration */
}
/* Create four equal columns that floats next to each other */
.column4 {
    /* float: left; */
    width: 25%;
    padding: 5px;
    /* height: 300px; /* Should be removed. Only for demonstration */
}

/* Clear floats after the columns */
.row:after {
    float: left;
    content: "";
    display: table;
    clear: both;
}
</style>

</head>


<body>
<!-- Navigation panel of all structures on top left of window -->
<div class="w3-sidebar w3-bar-block w3-light-grey w3-card" style="width:14%">
  <a href="#" class="w3-bar-item w3-button">Home</a> 
  <a href="#StructureReport" class="w3-bar-item w3-button">Structure Report</a> 
{% for structure in calculation_results.structure_graphics_dicts%} {#  structure_report_df_appended.iterrows() %} #}
  <a href="#{{ structure.html_div_id }}" class="w3-bar-item w3-button">
	  {{ structure.structure_id}} {{structure.mers}}</a>
{% endfor %}

<!-- show a doman graphic at top of main section -->

</div>
<div style="margin-left:15%">
<!-- This is where the domain graphic is targetted with canvas, by underlying routines -->
<table class="w3-table" style="width:50%">
	<tr>
		<th style="text-align:right">Date</th>
		<td>{{ today_date }}</td>
	</tr>
	<tr>
		<th style="text-align:right">Case</th>
		<td><a href="../index.html">{{ calculation_results.project }}</a></td>
	</tr>
	<tr>
		<th style="text-align:right">Gene</th>
		<td><a href="https://ghr.nlm.nih.gov/gene/{{ calculation_results.gene }}">{{ calculation_results.gene }}</a></td>
	</tr>
	<tr>
		<th style="text-align:right">Protein</th>
		<td><a href="https://www.uniprot.org/uniprot/{{ calculation_results.unp }}">{{ calculation_results.unp }}</a></td>
	</tr>
	<tr>
		<th style="text-align:right">Variant</th>
		<td>{{ calculation_results.variant }}</td>
	</tr>
</table>
<iframe  height="100%" width="100%" style="width:100%; height:100%; border: none; overflow:hidden; " src="{{ pfamGraphicsIframe_fname }}" onload="resizeIframe(this)" scrolling="no">
</iframe>
<!-- End Domain Graphic area -->


<!-- Present a table of all structures analyzed for the transcript -->
<hr id="StructureReport">
<div class="container">
    <h3>Structure Summary</h3>
<table cellspacing="1" class="tablesorter">
    <thead>
        <tr>
            <th>Method</th>
            <th>Structure Id</th>
            <th>Chain</th>
            <th>PDB<br />Template</th>
            <th>PDB<br />Resolution</th>
            <th>Template<br>Identity</th>
            <th>Transcript<br>Identity</th>
            <th class="sorter-false">AF Global<br>Quality</th>
            <th class="sorter-false">AF Residue<br>Confidence</th>
            <th>Seq<br />Start</th>
            <th>Seq<br />End</th>
            <th>ddG<br />Monomer</th>
            <th>ddG<br />Cartesian</th>
            <th>Pathprox<br>{{ config_pathprox_dict['disease1_variant_short_description'] }}</th>
            <th>Pathprox<br />{{ config_pathprox_dict['disease1_variant_short_description'] }} AOC</th>
            <th>Pathprox<br />{{ config_pathprox_dict['disease2_variant_short_description'] }}</th>
            <th>Pathprox<br />{{ config_pathprox_dict['disease2_variant_short_description'] }} AOC</th>
            <th>PDB<br />Pos</th>
        </tr>
    </thead>
    <tbody>
            {% for structure in calculation_results.structure_graphics_dicts%} {#  structure_report_df_appended.iterrows() %} #}
    <tr>
        <td>{{ structure.method }}</td>
        <td><A href="{{structure.structure_url}}">{{ structure.structure_id }}</A></td>
        <td>{{ structure.chain_id }}</td>
        <td>{{ structure.pdb_template }}</td>
        <td>{{ structure.resolution }}</td>
        <td>{{ structure.template_identity }}</td>
        <td>{{ structure.perc_identity }}</td>
        <td>{{ structure.alphafold_global_TMscore }}</td>
        <td>{{ structure.alphafold_residue_pLDDT }}</td>


        <td>{{ structure.trans_first_aligned }}</td>
        <td>{{ structure.trans_last_aligned }}</td>
        <td>{{ structure.ddG_monomer }}</td>
        <td>{{ structure.ddG_cartesian }}</td>
        <td>{{ structure.disease1_pathprox }}</td>
        <td>{{ structure.disease1_pr_auc }}</td>
        <td>{{ structure.disease2_pathprox }}</td>
        <td>{{ structure.disease2_pr_auc }}</td>
        <td>{{ structure.pdbmut }}</td>

        {#
        structure_id 	structure_url 	struct_filename 	mers 	pdb_template 	template_identity 	template_chain 	chain_id 	biounit 	method
        #}
    </tr>
            {% endfor %}
    </tbody>
</table>
</div>  <!-- End table of all structures at top of report -->



<!-- Out put structure graphics, and details now for Each structure. -->
	{% for structure in calculation_results.structure_graphics_dicts %}
    	<p style="page-break-before: always"></p>

        <hr id="{{ structure['html_div_id'] }}">
        <div class="container">
      	<div class="header">
                <h3 >{{ structure['structure_id'] }}.{{ structure['chain_id'] }} {{ structure['mers'] }}  - Structure Detail</h3>
      	</div>

    <!-- PFirst restate the line of structure details from the overall structure table -->
	<table cellspacing="1" class="tablesorter">
		<thead>
			<tr>
				<th class="sorter-false">Method</th>
				<th class="sorter-false">Structure Id</th>
				<th class="sorter-false">Chain</th>
				<th class="sorter-false">PDB<br/>Template</th>
				<th class="sorter-false">PDB<br /> Resolution</th>
				<th class="sorter-false">Template<br>Identity</th>
				<th class="sorter-false">Transcript<br>Identity</th>
                <th class="sorter-false">AF Global<br>Quality</th>
                <th lass="sorter-false">AF Residue<br>Confidence</th>

				<th class="sorter-false">Seq<br />Start</th>
				<th class="sorter-false">Seq<br />End</th>
				<th class="sorter-false">ddG<br />Monomer</th>
				<th class="sorter-false">ddG<br />Cartesian</th>
				<th class="sorter-false">Pathprox<br>{{ config_pathprox_dict['disease1_variant_short_description'] }}</th>
				<th class="sorter-false">Pathprox<br />{{ config_pathprox_dict['disease1_variant_short_description'] }} AOC</th>
				<th class="sorter-false">Pathprox<br />{{ config_pathprox_dict['disease2_variant_short_description'] }}</th>
				<th class="sorter-false">Pathprox<br />{{ config_pathprox_dict['disease2_variant_short_description'] }} AOC</th>
				<th class="sorter-false">PDB<br />Pos</th>
			</tr>
		</thead>
		<tbody>
		<tr>
			<td>{{ structure.method }}</td>
			<td><A href="{{structure.structure_url}}">{{ structure.structure_id }}</A></td>
			<td>{{ structure.chain_id }}</td>
			<td>{{ structure.pdb_template }}</td>
			<td>{{ structure.resolution }}</td>
			<td>{{ structure.template_identity }}</td>
			<td>{{ structure.perc_identity }}</td>
            <td>{{ structure.alphafold_global_TMscore }}</td>
            <td>{{ structure.alphafold_residue_pLDDT }}</td>

			<td>{{ structure.trans_first_aligned }}</td>
			<td>{{ structure.trans_last_aligned }}</td>
			<td>{{ structure.ddG_monomer }}</td>
			<td>{{ structure.ddG_cartesian }}</td>
			<td>{{ structure.disease1_pathprox }}</td>
			<td>{{ structure.disease1_pr_auc }}</td>
			<td>{{ structure.disease2_pathprox }}</td>
			<td>{{ structure.disease2_pr_auc }}</td>
			<td>{{ structure.pdbmut }}</td>

			{#
			structure_id 	structure_url 	struct_filename 	mers 	pdb_template 	template_identity 	template_chain 	chain_id 	biounit 	method
			#}
		</tr>
		</tbody>
	</table>

        {%  if structure['pathprox_disease1_results'] %}
			{# Now show the two pathprox results tables #}
        <!-- Structure Pathprox disease 1 table -->
        <h4>{{ config_pathprox_dict['disease1_variant_short_description'] }}
			Results - {{ structure.structure_id }}.{{ structure.chain_id }}</h4>
		<table cellspacing="1" class="tablesorter">
		  <thead>
			<tr>
			  <th class="sorter-false">Neutral Kz</th>
			  <th class="sorter-false">Neutral Kt</th>
			  <th class="sorter-false">Pathogenic Kz</th>
			  <th class="sorter-false">Pathogenic Kt</th>
			  <th class="sorter-false">ROC AUC</th>
			  <th class="sorter-false">PR AUC</th>
			  {# <th class="sorter-false">Mutation</th> #}
			  <th class="sorter-false">PathProx</th>
			  <th class="sorter-false">Neutral Constraint</th>
			  <th class="sorter-false">Pathogenic Constraint</th>
			</tr>
		  </thead>
		  <tbody>
			<tr>
			  <td>{{ "%.2f"|format(structure['pathprox_disease1_results']['Kz_neut']|float) }}</td>
			  <td>{{ "%.2f"|format(structure['pathprox_disease1_results']['Kt_neut']|float) }}</td>
			  <td>{{ "%.2f"|format(structure['pathprox_disease1_results']['Kz_path']|float) }}</td>
			  <td>{{ "%.2f"|format(structure['pathprox_disease1_results']['Kt_path']|float) }}</td>
			  <td>{{ "%.2f"|format(structure['pathprox_disease1_results']['roc_auc']|float) }}</td>
			  <td>{{ "%.2f"|format(structure['pathprox_disease1_results']['pr_auc']|float) }}</td>
		      {# <td>{{ structure['mutation'] }}</td> #}
		      <td>{{ "%.2f"|format(structure['pathprox_disease1_results']['pathprox']|float) }}</td>
		      <td>{{ "%.2f"|format(structure['pathprox_disease1_results']['neutcon']|float) }}</td>
		      <td>{{ "%.2f"|format(structure['pathprox_disease1_results']['pathcon']|float) }}</td>
			</tr>
		  </tbody>
		</table>
        {%  endif %}
			
        {%  if structure['pathprox_disease2_results'] %}
		{# Pathprox result set #2 #}
        <!-- Structure Pathprox disease 2 table -->

        <h4>{{ config_pathprox_dict['disease2_variant_short_description'] }}
            Results - {{ structure.structure_id }}.{{ structure.chain_id }}</h4>
		<table cellspacing="1" class="tablesorter">
		  <thead>
			<tr>
			  <th class="sorter-false">Neutral Kz</th>
			  <th class="sorter-false">Neutral Kt</th>
			  <th class="sorter-false">Pathogenic Kz</th>
			  <th class="sorter-false">Pathogenic Kt</th>
			  <th class="sorter-false">ROC AUC</th>
			  <th class="sorter-false">PR AUC</th>
			  <th class="sorter-false">PathProx</th>
			  <th class="sorter-false">Neutral Constraint</th>
			  <th class="sorter-false">Pathogenic Constraint</th>
			</tr>
		  </thead>
		  <tbody>
			<tr>
			  <td>{{ "%.2f"|format(structure['pathprox_disease2_results']['Kz_neut']|float) }}</td>
			  <td>{{ "%.2f"|format(structure['pathprox_disease2_results']['Kt_neut']|float) }}</td>
			  <td>{{ "%.2f"|format(structure['pathprox_disease2_results']['Kz_path']|float) }}</td>
			  <td>{{ "%.2f"|format(structure['pathprox_disease2_results']['Kt_path']|float) }}</td>
			  <td>{{ "%.2f"|format(structure['pathprox_disease2_results']['roc_auc']|float) }}</td>
			  <td>{{ "%.2f"|format(structure['pathprox_disease2_results']['pr_auc']|float) }}</td>
		      <td>{{ "%.2f"|format(structure['pathprox_disease2_results']['pathprox']|float) }}</td>
		      <td>{{ "%.2f"|format(structure['pathprox_disease2_results']['neutcon']|float) }}</td>
		      <td>{{ "%.2f"|format(structure['pathprox_disease2_results']['pathcon']|float) }}</td>
			</tr>
		  </tbody>
		</table>
        {%  endif %}

                  </script>
                <!-- NGL Viewer and some static  Plots -->
        <div class="container">
                <h4>Location of {{ calculation_results.variant }} in {{ structure.structure_id }}.{{ structure.chain_id }} (NGL Viewer)</h4>

<div style="height: 800px; width: 85%; border:2px solid black;">
        <iframe style="height: 100%"
                style="width:100%; height:100%; border: none; overflow:hidden;"
                src="{{ structure.html_div_id + '_viewer.html' }}"></iframe>
</div>
        </div>

                <!-- {{ config_pathprox_dict['neutral_variant_short_description'] }}, {{ config_pathprox_dict['disease1_variant_short_description'] }}, {{ config_pathprox_dict['disease2_variant_short_description'] }} Variant Mapping Plots -->
                <div class="wrapper main-box">
                        <h4>Variant Distributions</h4>

                        <div class="column3">
                                <h5>{{structure['ngl_neutral_residue_count'] if structure['ngl_neutral_residue_count']}} {{ config_pathprox_dict['neutral_variant_short_description'] }}</h5>
                                <img src="{{ structure['pathprox_disease1_results']['neutral_variants_png'] }}" class="img" style="width:100%;height:auto">
                        </div>
                        <div class="column3">
                                <h5>{{structure['ngl_disease1_residue_count'] if structure['ngl_disease1_residue_count']}} {{ config_pathprox_dict['disease1_variant_short_description'] }}</h5>
                                {% if structure['pathprox_disease1_results']['pathogenic_variants_png'] %}
                                        <img src="{{ structure['pathprox_disease1_results']['pathogenic_variants_png'] }}"  class="img" style="width:100%;height:auto">
                                {% else %}
                                        No {{ config_pathprox_dict['disease1_variant_short_description'] }} Variants map to this structure
                                {% endif %}
                        </div>
                        <div class="column3">
                                <h5>{{structure['ngl_disease2_residue_count'] if structure['ngl_disease2_residue_count']}} {{ config_pathprox_dict['disease2_variant_short_description'] }}</h5>
                                {% if structure['pathprox_disease2_results']['pathogenic_variants_png'] %}
                                        <img src="{{ structure['pathprox_disease2_results']['pathogenic_variants_png'] }}"  class="img" style="width:100%;height:auto">
                                {% else %}
                                        No {{ config_pathprox_dict['disease2_variant_short_description'] }} Variants map to this structure
                                {% endif %}
                        </div>
                </div>


                <div class="wrapper main-box">
                        <h4>Ripley's Univariate K</h4>
                        <div class="column3">
                                <h5>{{ config_pathprox_dict['neutral_variant_short_description'] }}</h5>
                                {% if structure['pathprox_disease1_results']['neutral_K_plot_png'] %}
                                        <img src="{{ structure['pathprox_disease1_results']['neutral_K_plot_png']}}" class="img" style="width:100%;height:auto">
                                {% else %}
                                        Ripley's Univariate K graphic was not created for {{ config_pathprox_dict['neutral_variant_short_description'] }}
                                {% endif %}
                        </div>
                        <div class="column3">
                                <h5>{{ config_pathprox_dict['disease1_variant_short_description'] }}</h5>
                                {% if structure['pathprox_disease1_results']['pathogenic_K_plot_png'] %}
                                        <img src="{{ structure['pathprox_disease1_results']['pathogenic_K_plot_png']}}" class="img" style="width:100%;height:auto">
                                {% else %}
                                        Ripley's Univariate K graphic was not created for {{ config_pathprox_dict['disease1_variant_short_description'] }}
                                {% endif %}
                        </div>
                        <div class="column3">
                                <h5>{{ config_pathprox_dict['disease2_variant_short_description'] }}</h5>
                                {% if structure['pathprox_disease2_results']['pathogenic_K_plot_png'] %}
                                        <img src="{{ structure['pathprox_disease2_results']['pathogenic_K_plot_png']}}" class="img" style="width:100%;height:auto">
                                {% else %}
                                        Ripley's Univariate K graphic was not created for {{ config_pathprox_dict['disease2_variant_short_description'] }}
                                {% endif %}
                        </div>
                </div>


                <!-- {{ config_pathprox_dict['disease1_variant_short_description'] }}-{{ config_pathprox_dict['neutral_variant_short_description'] }}, {{ config_pathprox_dict['disease2_variant_short_description'] }}-{{ config_pathprox_dict['neutral_variant_short_description'] }} Bivariate D Plots -->
                <div class="wrapper main-box">
                        <h4>Ripley's Bivariate D</h4>
                        <div class="column3">
                                <h5>{{ config_pathprox_dict['disease1_variant_short_description'] }} - {{ config_pathprox_dict['neutral_variant_short_description'] }}</h5>
                                {% if structure['pathprox_disease1_results']['D_plot_png'] %}
                                        <img src="{{ structure['pathprox_disease1_results']['D_plot_png']}}" class="img" style="width:100%;height:auto">
                                {% else %}
                                        {{ config_pathprox_dict['disease1_variant_short_description'] }}-{{ config_pathprox_dict['neutral_variant_short_description'] }} Bivariate D plot was not created
                                {% endif %}
                        </div>
                        <div class="column3">
                                <h5>{{ config_pathprox_dict['disease2_variant_short_description'] }} - {{ config_pathprox_dict['neutral_variant_short_description'] }}</h5>
                                {% if structure['pathprox_disease2_results']['D_plot_png'] %}
                                        <img src="{{ structure['pathprox_disease2_results']['D_plot_png']}}" class="img" style="width:100%;height:auto">
                                {% else %}
                                        {{ config_pathprox_dict['disease2_variant_short_description'] }}-{{ config_pathprox_dict['neutral_variant_short_description'] }} Bivariate D plot was not created
                                {% endif %}
                        </div>
                </div>

       <div class="wrapper main-box">
                <!-- {{ config_pathprox_dict['disease1_variant_short_description'] }}-{{ config_pathprox_dict['neutral_variant_short_description'] }} PathProx, ROC, PR -->
                <h4>PathProx ({{ config_pathprox_dict['disease1_variant_short_description'] }} - {{ config_pathprox_dict['neutral_variant_short_description'] }}) Results</h4>
                <div class="column3">
                        <h5>PathProx</h5>
                                {% if structure['pathprox_disease1_results']['pp_plot_png'] %}
                                        <img src="{{ structure['pathprox_disease1_results']['pp_plot_png']}}" class="img" style="width:100%;height:auto">
                        {% else %}
                                No graphic was created
                        {% endif %}
                </div>
                <div class="column3">
                        <h5>PathProx {{ config_pathprox_dict['disease1_variant_short_description'] }} ROC</h5>
                                {% if structure['pathprox_disease1_results']['roc_plot_png'] %}
                                        <img src="{{ structure['pathprox_disease1_results']['roc_plot_png']}}" class="img" style="width:100%;height:auto">
                        {% else %}
                                No graphic was created
                        {% endif %}
                </div>
                <div class="column3">
                                <h5>PathProx {{ config_pathprox_dict['disease1_variant_short_description'] }} PR</h5>
                                {% if structure['pathprox_disease1_results']['pr_plot_png'] %}
                                        <img src="{{ structure['pathprox_disease1_results']['pr_plot_png']}}" class="img" style="width:100%;height:auto">
                        {% else %}
                                No graphic was created
                        {% endif %}
                </div>
       </div>

                <!-- {{ config_pathprox_dict['disease2_variant_short_description'] }}-{{ config_pathprox_dict['neutral_variant_short_description'] }} PathProx, ROC, PR -->
        <div class="wrapper main-box">
                        <h4>PathProx ({{ config_pathprox_dict['disease2_variant_short_description'] }} - {{ config_pathprox_dict['neutral_variant_short_description'] }}) Results</h4>
                <div class="column3">
                        <h5>PathProx</h5>
                                {% if structure['pathprox_disease2_results']['pp_plot_png'] %}
                                        <img src="{{ structure['pathprox_disease2_results']['pp_plot_png']}}" class="img" style="width:100%;height:auto">
                        {% else %}
                                No graphic was created
                        {% endif %}
                </div>
                <div class="column3">
                        <h5>PathProx {{ config_pathprox_dict['disease2_variant_short_description'] }} ROC</h5>
                                {% if structure['pathprox_disease2_results']['roc_plot_png'] %}
                                        <img src="{{ structure['pathprox_disease2_results']['roc_plot_png']}}" class="img" style="width:100%;height:auto">
                        {% else %}
                                No graphic was created
                        {% endif %}
                </div>
                <div class="column3">
                                <h5>PathProx {{ config_pathprox_dict['disease2_variant_short_description'] }} PR</h5>
                                {% if structure['pathprox_disease2_results']['pr_plot_png'] %}
                                        <img src="{{ structure['pathprox_disease2_results']['pr_plot_png']}}" class="img" style="width:100%;height:auto">
                        {% else %}
                                No graphic was created
                        {% endif %}
                </div>
        </div>

        </div>



			{% endfor %}     {# for loop over all structures #}






</div>
</body>
</html>
