<html><head><meta content="text/html; charset=UTF-8" http-equiv="content-type"><style type="text/css">.lst-kix_xniqrz2fvjl-0>li:before{content:"\0025cf   "}.lst-kix_xniqrz2fvjl-1>li:before{content:"\0025cb   "}.lst-kix_rtv2mok54gp3-1>li{counter-increment:lst-ctn-kix_rtv2mok54gp3-1}.lst-kix_xniqrz2fvjl-2>li:before{content:"\0025a0   "}ul.lst-kix_unolgk77u2yq-0{list-style-type:none}ul.lst-kix_unolgk77u2yq-1{list-style-type:none}.lst-kix_rtv2mok54gp3-7>li{counter-increment:lst-ctn-kix_rtv2mok54gp3-7}.lst-kix_xniqrz2fvjl-7>li:before{content:"\0025cb   "}.lst-kix_xniqrz2fvjl-6>li:before{content:"\0025cf   "}ol.lst-kix_rtv2mok54gp3-2.start{counter-reset:lst-ctn-kix_rtv2mok54gp3-2 0}ol.lst-kix_rtv2mok54gp3-8.start{counter-reset:lst-ctn-kix_rtv2mok54gp3-8 0}.lst-kix_xniqrz2fvjl-4>li:before{content:"\0025cb   "}ul.lst-kix_unolgk77u2yq-4{list-style-type:none}ul.lst-kix_unolgk77u2yq-5{list-style-type:none}.lst-kix_rtv2mok54gp3-2>li{counter-increment:lst-ctn-kix_rtv2mok54gp3-2}.lst-kix_xniqrz2fvjl-3>li:before{content:"\0025cf   "}.lst-kix_xniqrz2fvjl-5>li:before{content:"\0025a0   "}ul.lst-kix_unolgk77u2yq-2{list-style-type:none}ul.lst-kix_unolgk77u2yq-3{list-style-type:none}ul.lst-kix_unolgk77u2yq-8{list-style-type:none}.lst-kix_rtv2mok54gp3-8>li{counter-increment:lst-ctn-kix_rtv2mok54gp3-8}ul.lst-kix_unolgk77u2yq-6{list-style-type:none}ul.lst-kix_unolgk77u2yq-7{list-style-type:none}ol.lst-kix_rtv2mok54gp3-5.start{counter-reset:lst-ctn-kix_rtv2mok54gp3-5 0}ol.lst-kix_rtv2mok54gp3-6.start{counter-reset:lst-ctn-kix_rtv2mok54gp3-6 0}.lst-kix_unolgk77u2yq-8>li:before{content:"\0025a0   "}.lst-kix_unolgk77u2yq-7>li:before{content:"\0025cb   "}.lst-kix_rtv2mok54gp3-7>li:before{content:"" counter(lst-ctn-kix_rtv2mok54gp3-7,lower-latin) ". "}.lst-kix_rtv2mok54gp3-6>li:before{content:"" counter(lst-ctn-kix_rtv2mok54gp3-6,decimal) ". "}.lst-kix_rtv2mok54gp3-8>li:before{content:"" counter(lst-ctn-kix_rtv2mok54gp3-8,lower-roman) ". "}.lst-kix_xniqrz2fvjl-8>li:before{content:"\0025a0   "}.lst-kix_rtv2mok54gp3-1>li:before{content:"" counter(lst-ctn-kix_rtv2mok54gp3-1,lower-latin) ") "}.lst-kix_rtv2mok54gp3-6>li{counter-increment:lst-ctn-kix_rtv2mok54gp3-6}ul.lst-kix_xniqrz2fvjl-0{list-style-type:none}.lst-kix_rtv2mok54gp3-3>li:before{content:"(" counter(lst-ctn-kix_rtv2mok54gp3-3,decimal) ") "}.lst-kix_rtv2mok54gp3-3>li{counter-increment:lst-ctn-kix_rtv2mok54gp3-3}ul.lst-kix_xniqrz2fvjl-2{list-style-type:none}.lst-kix_rtv2mok54gp3-2>li:before{content:"" counter(lst-ctn-kix_rtv2mok54gp3-2,lower-roman) ") "}ul.lst-kix_xniqrz2fvjl-1{list-style-type:none}ul.lst-kix_xniqrz2fvjl-4{list-style-type:none}.lst-kix_rtv2mok54gp3-5>li:before{content:"(" counter(lst-ctn-kix_rtv2mok54gp3-5,lower-roman) ") "}ul.lst-kix_xniqrz2fvjl-3{list-style-type:none}ul.lst-kix_xniqrz2fvjl-6{list-style-type:none}ul.lst-kix_xniqrz2fvjl-5{list-style-type:none}ul.lst-kix_xniqrz2fvjl-8{list-style-type:none}.lst-kix_rtv2mok54gp3-0>li{counter-increment:lst-ctn-kix_rtv2mok54gp3-0}ul.lst-kix_xniqrz2fvjl-7{list-style-type:none}.lst-kix_rtv2mok54gp3-4>li:before{content:"(" counter(lst-ctn-kix_rtv2mok54gp3-4,lower-latin) ") "}.lst-kix_h9zr6ilvhp1k-2>li:before{content:"\0025a0   "}ol.lst-kix_rtv2mok54gp3-7{list-style-type:none}ol.lst-kix_rtv2mok54gp3-0.start{counter-reset:lst-ctn-kix_rtv2mok54gp3-0 0}ol.lst-kix_rtv2mok54gp3-8{list-style-type:none}.lst-kix_h9zr6ilvhp1k-3>li:before{content:"\0025cf   "}.lst-kix_h9zr6ilvhp1k-0>li:before{content:"\0025cf   "}.lst-kix_h9zr6ilvhp1k-8>li:before{content:"\0025a0   "}.lst-kix_rtv2mok54gp3-4>li{counter-increment:lst-ctn-kix_rtv2mok54gp3-4}.lst-kix_h9zr6ilvhp1k-1>li:before{content:"\0025cb   "}.lst-kix_rtv2mok54gp3-0>li:before{content:"" counter(lst-ctn-kix_rtv2mok54gp3-0,decimal) ") "}ol.lst-kix_rtv2mok54gp3-7.start{counter-reset:lst-ctn-kix_rtv2mok54gp3-7 0}.lst-kix_unolgk77u2yq-1>li:before{content:"\0025cb   "}ol.lst-kix_rtv2mok54gp3-3.start{counter-reset:lst-ctn-kix_rtv2mok54gp3-3 0}.lst-kix_unolgk77u2yq-2>li:before{content:"\0025a0   "}.lst-kix_unolgk77u2yq-3>li:before{content:"\0025cf   "}.lst-kix_unolgk77u2yq-5>li:before{content:"\0025a0   "}.lst-kix_rtv2mok54gp3-5>li{counter-increment:lst-ctn-kix_rtv2mok54gp3-5}.lst-kix_unolgk77u2yq-4>li:before{content:"\0025cb   "}.lst-kix_unolgk77u2yq-6>li:before{content:"\0025cf   "}ul.lst-kix_h9zr6ilvhp1k-7{list-style-type:none}ul.lst-kix_h9zr6ilvhp1k-8{list-style-type:none}ul.lst-kix_h9zr6ilvhp1k-5{list-style-type:none}ul.lst-kix_h9zr6ilvhp1k-6{list-style-type:none}ul.lst-kix_h9zr6ilvhp1k-3{list-style-type:none}ul.lst-kix_h9zr6ilvhp1k-4{list-style-type:none}ul.lst-kix_h9zr6ilvhp1k-1{list-style-type:none}ul.lst-kix_h9zr6ilvhp1k-2{list-style-type:none}ul.lst-kix_h9zr6ilvhp1k-0{list-style-type:none}li.li-bullet-0:before{margin-left:-18pt;white-space:nowrap;display:inline-block;min-width:18pt}.lst-kix_unolgk77u2yq-0>li:before{content:"\0025cf   "}.lst-kix_h9zr6ilvhp1k-7>li:before{content:"\0025cb   "}ol.lst-kix_rtv2mok54gp3-4.start{counter-reset:lst-ctn-kix_rtv2mok54gp3-4 0}.lst-kix_h9zr6ilvhp1k-6>li:before{content:"\0025cf   "}ol.lst-kix_rtv2mok54gp3-1{list-style-type:none}ol.lst-kix_rtv2mok54gp3-2{list-style-type:none}ol.lst-kix_rtv2mok54gp3-1.start{counter-reset:lst-ctn-kix_rtv2mok54gp3-1 0}ol.lst-kix_rtv2mok54gp3-0{list-style-type:none}.lst-kix_h9zr6ilvhp1k-4>li:before{content:"\0025cb   "}ol.lst-kix_rtv2mok54gp3-5{list-style-type:none}ol.lst-kix_rtv2mok54gp3-6{list-style-type:none}.lst-kix_h9zr6ilvhp1k-5>li:before{content:"\0025a0   "}ol.lst-kix_rtv2mok54gp3-3{list-style-type:none}ol.lst-kix_rtv2mok54gp3-4{list-style-type:none}ol{margin:0;padding:0}table td,table th{padding:0}.c0{padding-top:18pt;padding-bottom:6pt;line-height:1.15;page-break-after:avoid;orphans:2;widows:2;text-align:left;height:16pt}.c1{margin-left:36pt;padding-top:0pt;padding-left:0pt;padding-bottom:0pt;line-height:1.15;orphans:2;widows:2;text-align:left}.c11{color:#000000;font-weight:400;text-decoration:none;vertical-align:baseline;font-size:16pt;font-family:"Arial";font-style:normal}.c14{color:#434343;font-weight:400;text-decoration:none;vertical-align:baseline;font-size:14pt;font-family:"Arial";font-style:normal}.c9{color:#000000;font-weight:400;text-decoration:none;vertical-align:baseline;font-size:20pt;font-family:"Arial";font-style:normal}.c20{color:#000000;font-weight:400;text-decoration:none;vertical-align:baseline;font-size:26pt;font-family:"Arial";font-style:normal}.c16{padding-top:16pt;padding-bottom:4pt;line-height:1.15;page-break-after:avoid;orphans:2;widows:2;text-align:left}.c4{padding-top:18pt;padding-bottom:6pt;line-height:1.15;page-break-after:avoid;orphans:2;widows:2;text-align:left}.c15{padding-top:20pt;padding-bottom:6pt;line-height:1.15;page-break-after:avoid;orphans:2;widows:2;text-align:left}.c13{padding-top:0pt;padding-bottom:3pt;line-height:1.15;page-break-after:avoid;orphans:2;widows:2;text-align:left}.c3{color:#000000;font-weight:400;text-decoration:none;vertical-align:baseline;font-size:11pt;font-family:"Arial";font-style:normal}.c2{padding-top:0pt;padding-bottom:0pt;line-height:1.15;orphans:2;widows:2;text-align:left}.c10{padding-top:12pt;padding-bottom:12pt;line-height:1.15;orphans:2;widows:2;text-align:left}.c7{padding-top:0pt;padding-bottom:0pt;line-height:1.0;text-align:left}.c19{background-color:#ffffff;max-width:468pt;padding:72pt 72pt 72pt 72pt}.c5{margin-left:72pt;height:11pt}.c18{border:1px solid black;margin:5px}.c17{margin-left:72pt;padding-left:0pt}.c12{padding:0;margin:0}.c6{height:11pt}.c8{font-weight:700}.title{padding-top:0pt;color:#000000;font-size:26pt;padding-bottom:3pt;font-family:"Arial";line-height:1.15;page-break-after:avoid;orphans:2;widows:2;text-align:left}.subtitle{padding-top:0pt;color:#666666;font-size:15pt;padding-bottom:16pt;font-family:"Arial";line-height:1.15;page-break-after:avoid;orphans:2;widows:2;text-align:left}li{color:#000000;font-size:11pt;font-family:"Arial"}p{margin:0;color:#000000;font-size:11pt;font-family:"Arial"}h1{padding-top:20pt;color:#000000;font-size:20pt;padding-bottom:6pt;font-family:"Arial";line-height:1.15;page-break-after:avoid;orphans:2;widows:2;text-align:left}h2{padding-top:18pt;color:#000000;font-size:16pt;padding-bottom:6pt;font-family:"Arial";line-height:1.15;page-break-after:avoid;orphans:2;widows:2;text-align:left}h3{padding-top:16pt;color:#434343;font-size:14pt;padding-bottom:4pt;font-family:"Arial";line-height:1.15;page-break-after:avoid;orphans:2;widows:2;text-align:left}h4{padding-top:14pt;color:#666666;font-size:12pt;padding-bottom:4pt;font-family:"Arial";line-height:1.15;page-break-after:avoid;orphans:2;widows:2;text-align:left}h5{padding-top:12pt;color:#666666;font-size:11pt;padding-bottom:4pt;font-family:"Arial";line-height:1.15;page-break-after:avoid;orphans:2;widows:2;text-align:left}h6{padding-top:12pt;color:#666666;font-size:11pt;padding-bottom:4pt;font-family:"Arial";line-height:1.15;page-break-after:avoid;font-style:italic;orphans:2;widows:2;text-align:left}</style></head><body class="c19 doc-content"><p class="c2"><span class="c3">[top of page will have same header as https://structbio.vanderbilt.edu/~mothcw/VUStructDocs/ ]</span></p><p class="c2 c6"><span class="c3"></span></p><p class="c13 title" id="h.wyoagvss0z8h"><span class="c20">VUStruct tutorial</span></p><p class="c2"><span class="c3">Documentation currently located at: https://structbio.vanderbilt.edu/~mothcw/VUStructDocs/</span></p><h1 class="c15" id="h.a370vdz60z8r"><span class="c9">Background</span></h1><p class="c2"><span class="c3">Many tools have been developed to predict the effect of missense variants in the genome. VUStruct is provided to additionally analyze factors that are often excluded by other tools, such as thermodynamic destabilization of the resulting protein structure, 3D colocalization of the variant with clusters of pathogenic mutations in the same protein, potential digenic effects arising from pairwise interactions of variants in different genes, and other factors described below. For details about how the various methods work, please read the pages on Calculation Details:</span></p><p class="c2"><span class="c3">https://structbio.vanderbilt.edu/~mothcw/VUStructDocs/CalcDetailsRosetta.html</span></p><p class="c2"><span class="c3">https://structbio.vanderbilt.edu/~mothcw/VUStructDocs/CalcDetailsPathProx.html</span></p><p class="c2"><span class="c3">&nbsp;&lt;Evolutionary Constraint&gt; </span></p><p class="c2"><span class="c3">&nbsp;&lt;Digenic Disease Prediction&gt;</span></p><p class="c2"><span class="c3">&nbsp;&lt;PPI Site Prediction&gt;</span></p><p class="c2"><span class="c3">&nbsp;&lt;PTM Prediction&gt;</span></p><p class="c2 c6"><span class="c3"></span></p><p class="c2"><span class="c3">Here you will start with an example list of variants as input. These variants come from the sequencing results of a fictional case for diagnosis of an unknown disease. This tutorial will walk you through preparation and submission of the input data, and the interpretation of the results generated by the system.</span></p><p class="c2 c6"><span class="c3"></span></p><h1 class="c15" id="h.lkqhrh14kef7"><span class="c9">Inputs</span></h1><p class="c2"><span class="c3">VUStruct is designed to accept a list of candidate variants for a single case. Each case typically includes 5-50 variants selected from WES or WGS of the proband. This selection removes variants of unreliable quality, variants too common to be associated with the proband&rsquo;s phenotype, variants associated with conditions unrelated to the patient&rsquo;s condition, and known benign variants. In the UDN workflow, this step is usually performed by experts in clinical genetics, using standard software tools like VEP and databases of biomedical literature like OMIM. </span></p><p class="c2 c6"><span class="c3"></span></p><p class="c2"><span class="c3">The length of the input lists for this webserver is limited to 100 variants, because the ddG calculations are computationally intensive. But please contact us if you want to submit a larger set.</span></p><p class="c2 c6"><span class="c3"></span></p><p class="c2"><span class="c3">Use only de-identified data. You will be given a private link to access the results.</span></p><p class="c2 c6"><span class="c3"></span></p><p class="c2"><span class="c3">Include inheritance information if you have it. Typically the proband&rsquo;s parents are included in the WES and WGS testing, to ascertain the inheritance of the variants, and match it with the segregation of the phenotype. Sometimes additional relatives are included, e.g. siblings who are affected or unaffected. </span></p><p class="c2 c6"><span class="c3"></span></p><p class="c2"><span>For this example,</span><sup><a href="#cmnt1" id="cmnt_ref1">[a]</a></sup><span class="c3">&nbsp;we provide the input data in two formats. You can use either one- they contain equivalent information.</span></p><p class="c2 c6"><span class="c3"></span></p><p class="c2"><span class="c3">Might start with a list derived (but modified) from https://structbio.vanderbilt.edu/~sheehajh/psb/VUDP_017_JD/</span></p><p class="c2 c6"><span class="c3"></span></p><h3 class="c16" id="h.df8kt96vx41f"><span class="c14">Input format 1:</span></h3><p class="c2"><span class="c3">Two files: the VCF file, and the inheritance information file.</span></p><h3 class="c16" id="h.llvouh410d0l"><span class="c14">Input format 2:</span></h3><p class="c2"><span class="c3">A custom MSExcel worksheet which includes the variants and the inheritance together.</span></p><p class="c2 c6"><span class="c3"></span></p><h1 class="c15" id="h.qou5kyf0pa59"><span class="c9">Launching the calculation</span></h1><p class="c2"><span class="c3">Start at the main VUStruct submission page, which is located here:</span></p><p class="c2"><span class="c3">&nbsp; &nbsp; &nbsp; https://meilerlab.org/VUStruct</span></p><p class="c2"><span class="c3">[fig00.png]</span></p><p class="c2"><span style="overflow: hidden; display: inline-block; margin: 0.00px 0.00px; border: 0.00px solid #000000; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px); width: 1289.00px; height: 1045.00px;"><img alt="" src="images/image7.png" style="width: 1289.00px; height: 1045.00px; margin-left: 0.00px; margin-top: 0.00px; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px);" title=""></span></p><p class="c2 c6"><span class="c3"></span></p><p class="c2"><span class="c3">Provide a Label for this run in the &ldquo;Case Label&rdquo; field.</span></p><p class="c2 c6"><span class="c3"></span></p><p class="c2"><span class="c3">Select a file from your hard drive, or copy-and-paste the VCF file information into the form. We provide an example input file for this tutorial, so that you have an input that is known to work correctly. You can use either format:</span></p><p class="c2 c6"><span class="c3"></span></p><ol class="c12 lst-kix_rtv2mok54gp3-0 start" start="1"><li class="c1 li-bullet-0"><span class="c3">Example input file in .xlsx format (see link above)</span></li><li class="c1 li-bullet-0"><span class="c3">Example input file in VCF format (see link above)</span></li></ol><p class="c2 c6"><span class="c3"></span></p><p class="c2"><span class="c3">Press the green Submit button at the bottom of the page.</span></p><p class="c2 c6"><span class="c3"></span></p><p class="c2"><span class="c3">You will receive a link to a page where the results will appear. You can bookmark that link.</span></p><p class="c2"><span class="c3">The duration of the calculation depends primarily on the size of the proteins. &nbsp;If the longest protein in your list is under 200 residues in length, the calculation can finish in as little as 4-6 hours. For proteins &gt;1000 residues in length, the calculation can take more than 3 days to finish.</span></p><p class="c2 c6"><span class="c3"></span></p><h1 class="c15" id="h.4ynqx59ocipu"><span class="c9">Problem Solving</span></h1><ul class="c12 lst-kix_xniqrz2fvjl-0 start"><li class="c1 li-bullet-0"><span class="c3">VUStruct currently works with the GRCh38 genome assembly.</span></li><li class="c1 li-bullet-0"><span class="c3">It is best practice to use the approved HGNC names for genes, rather than synonyms, but the chromosomal location is the fundamental starting point for the calculations.</span></li><li class="c1 li-bullet-0"><span class="c3">Most genes have multiple transcripts, corresponding to multiple protein isoforms. VUStruct starts from the genetic coordinates, and maps the variant you provide to each annotated transcript. If you run into a problem, try providing the variant in the context of the MANE-select default transcript. (https://useast.ensembl.org/info/genome/genebuild/mane.html) </span></li><li class="c1 li-bullet-0"><span class="c3">The software has been tested with files generated on Windows, Macintosh, and Linux. If you are using an unusual encoding, please contact us.</span></li></ul><p class="c2 c6"><span class="c3"></span></p><h1 class="c15" id="h.sp5eqzlrfxtp"><span class="c9">Interpreting Results</span></h1><h2 class="c4" id="h.iyht3tfg2fhx"><span class="c11">General</span></h2><p class="c2"><span class="c3">The results page for a calculation looks like this. The top table shows a summary of the predictions for all variants submitted.The DiGePred predictions are shown in a heat map linked from &ldquo;DiGePred Analysis,&rdquo; and summarized in the lower table.</span></p><p class="c2 c6"><span class="c3"></span></p><p class="c2"><span>Each line in the top table shows the variant information, and a </span><span class="c8">summary </span><span class="c3">of the PathProx and ddG calculations. Note that the summary does not give a full picture of the predictions, because it combines calculations performed on multiple isoforms, and multiple structural models for each variant. You should look at each one in detail as shown below:</span></p><p class="c2 c6"><span class="c3"></span></p><p class="c2"><span>You should first </span><span class="c8">expand </span><span class="c3">the rows for each variant of interest by clicking on the &ldquo;collapsable&rdquo; icon (looks like a greater-than sign) at the left of each row. Example highlighted in red box here [fig03.png]:</span></p><p class="c10"><span style="overflow: hidden; display: inline-block; margin: 0.00px 0.00px; border: 0.00px solid #000000; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px); width: 1170.00px; height: 1026.00px;"><img alt="" src="images/image9.png" style="width: 1170.00px; height: 1026.00px; margin-left: 0.00px; margin-top: 0.00px; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px);" title=""></span></p><p class="c2"><span>Once that row is expanded, you will see the </span><span class="c8">links </span><span>in the first column go to a &ldquo;Report&rdquo; for the variant as it appears in each curated transcript (and corresponding protein isoform). &nbsp;Example highlighted in red below [fig 05.png]. These are important links, because they will take you to the detailed prediction for the calculations. Again, you should not rely on the summary alone. Click on a &ldquo;</span><span class="c8">Report</span><span class="c3">&rdquo; link to see the results page for that variant.</span></p><p class="c2 c6"><span class="c3"></span></p><p class="c2"><span style="overflow: hidden; display: inline-block; margin: 0.00px 0.00px; border: 0.00px solid #000000; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px); width: 861.00px; height: 1026.00px;"><img alt="" src="images/image8.png" style="width: 861.00px; height: 1026.00px; margin-left: 0.00px; margin-top: 0.00px; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px);" title=""></span></p><p class="c2 c6"><span class="c3"></span></p><h2 class="c4" id="h.kcqez37b3zxx"><span class="c11">Report Pages</span></h2><p class="c2 c6"><span class="c3"></span></p><p class="c2"><span>Each </span><span class="c8">Report</span><span class="c3">&nbsp;link will take you to another page that is specific for that variant in one transcript/protein isoform. The calculations can be performed multiple times, depending on how many structural models are available for the protein (which include the variant position). &nbsp;An example Report page is shown below. Helpful features to note on this Report page are:</span></p><p class="c2 c5"><span class="c3"></span></p><ul class="c12 lst-kix_h9zr6ilvhp1k-0 start"><li class="c2 c17 li-bullet-0"><span class="c3">Summary information at the top, including links back to the main &ldquo;case&rdquo; page, the gene information at Medline+, and the protein information at Uniprot. </span></li><li class="c2 c17 li-bullet-0"><span class="c3">A graphical view of the protein sequence, with PFAM domains highlighted, and showing the location of the variant of interest. This quickly shows whether the variant falls in an annotated domain, between domains, or near the protein termini. This is a quick view; for greater detail, you can look at the Interpro site at EMBL-EBI.</span></li><li class="c2 c17 li-bullet-0"><span class="c3">Rate4Site score for this position in the protein (for details, see the documentation page.)</span></li><li class="c2 c17 li-bullet-0"><span class="c3">Cosmis score for this position in the protein (for details, see the documentation page.)</span></li><li class="c2 c17 li-bullet-0"><span class="c8">Structure Summary</span><span class="c3">&nbsp;table lists all structures used for the calculations. The PathProx and ddG calculations are run for each structure. There may be experimental structures from the PDB, including structures generated by X-ray crystallography, NMR solution spectroscopy, and CryoEM techniques. There also may be models gathered from the AlphaFold2 database (the names start with &ldquo;AF-&rdquo;). There may be models taken from ModBase at Salilab (the names start with the RefSeqID &ldquo;ENSP&hellip;&rdquo; ). &nbsp;And models from SWISS-MODEL repo at Expasy (the names start with the UniProt ID, e.g. Q13825). Some of these structures and models will contain the full-length protein, sometimes in multimeric states, others contain only domains or a fragment of the protein. The structures and models may vary widely in quality and relevance, so you will want to consider which one is the best for your purpose.</span></li><li class="c2 c17 li-bullet-0"><span>The left margin contains </span><span class="c8">shortcuts </span><span class="c3">to the section of the page that contains the results for each structure listed in the Structure Summary table. Each section contains details of the PathProx and ddG calculations. To be clear: the links in the table take you to the source of each structural model, whereas the shortcuts in the left sidebar take you to the results of the calculations lower down on this page. Click on one of the shortcuts to go to that section. (Highlighted in red here.)</span></li></ul><p class="c2"><span class="c3">[fig07.png]</span></p><p class="c2 c6"><span class="c3"></span></p><p class="c2"><span style="overflow: hidden; display: inline-block; margin: 0.00px 0.00px; border: 0.00px solid #000000; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px); width: 1236.00px; height: 1045.00px;"><img alt="" src="images/image3.png" style="width: 1236.00px; height: 1045.00px; margin-left: 0.00px; margin-top: 0.00px; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px);" title=""></span></p><p class="c2 c6"><span class="c3"></span></p><h2 class="c4" id="h.63xciopb5k4w"><span class="c11">Example section for a structural model</span></h2><p class="c2"><span class="c3">Each section contains results for one of the structural models. The results include: </span></p><p class="c2 c6"><span class="c3"></span></p><ul class="c12 lst-kix_unolgk77u2yq-0 start"><li class="c1 li-bullet-0"><span class="c3">PathProx calculations using variants from ClinVar as &ldquo;Pathogenic&rdquo;</span></li><li class="c1 li-bullet-0"><span class="c3">PathProx calculations using variants from COSMIC as &ldquo;Pathogenic&rdquo;</span></li><li class="c1 li-bullet-0"><span class="c3">a 3D display (using NGL viewer) of the variant in the structural context </span></li><li class="c1 li-bullet-0"><span class="c3">and 14 thumbnails detailing the variant distributions and statistics for the PathProx calculations</span></li></ul><p class="c2 c6"><span class="c3"></span></p><p class="c2"><span class="c3">Here are shown the results of calculations using the model from ModBase. [fig09.png] You can see that in this case, the ddG_monomer calculation gave a very high score, &gt;46 kcal/mol, in contrast with the ddG_cartesian calculation which did not give a score.We launch both protocols because sometimes one of them fails to complete, due to problems such as e.g. chainbreaks in the input structure. Below that, the PathProx score of 0.27 is shown under &ldquo;ClinVar Results,&rdquo; which quantifies the colocalization of the variant compared to distributions of pathogenic vs benign variants (for details and references, look back at the help page- &nbsp;https://structbio.vanderbilt.edu/~mothcw/VUStructDocs/CalcDetailsPathProx.html ).</span></p><p class="c2"><span class="c3">&nbsp;</span></p><p class="c2"><span style="overflow: hidden; display: inline-block; margin: 0.00px 0.00px; border: 0.00px solid #000000; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px); width: 1236.00px; height: 1045.00px;"><img alt="" src="images/image4.png" style="width: 1236.00px; height: 1045.00px; margin-left: 0.00px; margin-top: 0.00px; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px);" title=""></span></p><p class="c2 c6"><span class="c3"></span></p><h2 class="c4" id="h.o572noq647c1"><span class="c11">Structure display</span></h2><p class="c2"><span>The variant is shown in the context of the structural model using NGL viewer. This provides great flexibility in displaying the data, even permitting you to </span><span class="c8">save a figure</span><span class="c3">. For example, use the menu option Variants &rarr; ClinVar to show the locations of the pathogenic variants. Then use Colors &rarr; ClinVar PathProx to paint the ribbons with the corresponding scores at each residue. It becomes clear why this location scored highly, especially if you contrast that with the display of Variants &rarr; GnomAD. You can save your figure with File &rarr; &ldquo;Export Image&rdquo; &nbsp;[Fig10.png]</span></p><p class="c2 c6"><span class="c3"></span></p><p class="c2"><span style="overflow: hidden; display: inline-block; margin: 0.00px 0.00px; border: 0.00px solid #000000; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px); width: 889.00px; height: 687.00px;"><img alt="" src="images/image5.png" style="width: 889.00px; height: 687.00px; margin-left: 0.00px; margin-top: 0.00px; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px);" title=""></span></p><h2 class="c4" id="h.vx01lamrnsd9"><span class="c11">Thumbnails of PathProx Stats</span></h2><p class="c2"><span class="c3">The 3D viewer is followed by additional images showing the distributions of variants compared to the statistics of the null distribution used by PathProx</span></p><p class="c2 c6"><span class="c3"></span></p><p class="c2"><span class="c3">&nbsp;[Fig11.png]</span></p><p class="c2"><span style="overflow: hidden; display: inline-block; margin: 0.00px 0.00px; border: 0.00px solid #000000; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px); width: 1236.00px; height: 1045.00px;"><img alt="" src="images/image6.png" style="width: 1236.00px; height: 1045.00px; margin-left: 0.00px; margin-top: 0.00px; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px);" title=""></span></p><p class="c2 c6"><span class="c3"></span></p><p class="c2"><span>It can be useful to view the ROC curve and the PR curve for the PathProx calculations. The score itself can be more or less convincing, depending on the number and distribution of the pathogenic variants and the neutral variants. You can see in the PathProx Clinvar ROC below, that the classification confidence is suggestive, but not overwhelmingly strong. </span><span class="c8">Does that match your intuition</span><span class="c3">&nbsp;from looking at the distributions in the NGL viewer above? For a better understanding of spatial constraint, be sure to read Sively R.M HJHG 2018. https://pubmed.ncbi.nlm.nih.gov/29455857/</span></p><p class="c2 c6"><span class="c3"></span></p><p class="c2"><span class="c3">[fig_12.png]</span></p><p class="c2"><span style="overflow: hidden; display: inline-block; margin: 0.00px 0.00px; border: 0.00px solid #000000; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px); width: 1236.00px; height: 1045.00px;"><img alt="" src="images/image2.png" style="width: 1236.00px; height: 1045.00px; margin-left: 0.00px; margin-top: 0.00px; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px);" title=""></span></p><p class="c2 c6"><span class="c3"></span></p><p class="c2 c6"><span class="c3"></span></p><p class="c2 c6"><span class="c3"></span></p><h2 class="c0" id="h.lrv3d5vgocj"><span class="c11"></span></h2><p class="c2 c6"><span class="c3"></span></p><h2 class="c4" id="h.4srg04sdtu67"><span class="c11">DiGePred results</span></h2><p class="c2"><span class="c3">The genes you provide in your list are evaluated for the potential of digenic disease interactions. This is done using a predictor called DiGePred (https://pubmed.ncbi.nlm.nih.gov/34529933/ ), a random forest classifier for identifying candidate digenic disease gene pairs by features derived from biological networks, genomics, evolutionary history, and functional annotations. Go back to the &ldquo;case page&rdquo; that summarizes all the results for this VUStruct run. You&rsquo;ll see a summary table of the DiGePred scores. Just above that is the link to the detailed results (highlighted in red below):</span></p><p class="c2 c6"><span class="c3"></span></p><p class="c2 c6"><span class="c3"></span></p><p class="c2"><span class="c3">[fig_14.png]</span></p><p class="c2"><span style="overflow: hidden; display: inline-block; margin: 0.00px 0.00px; border: 0.00px solid #000000; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px); width: 829.00px; height: 1045.00px;"><img alt="" src="images/image1.png" style="width: 829.00px; height: 1045.00px; margin-left: 0.00px; margin-top: 0.00px; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px);" title=""></span></p><p class="c2 c6"><span class="c3"></span></p><p class="c2"><span class="c3">The digenic predictions are displayed as a heat map. Each cell represents the potential interactions between two of the genes in the list. The matrix is symmetric (redundant). The legend on the right hand side indicates the colors that the cell borders will be drawn, used to highlight the inheritance information (if provided with the input data). &nbsp;The tabs at the bottom are used to select subsets of the information used as support for the predicted interactions, and change the display of the heatmap. Moving the mouse cursor over individual cells displays a pop-up tip that summarizes the score, and the information used to support that score.</span></p><p class="c2 c6"><span class="c3"></span></p><p class="c2"><span class="c3">[fig_16.png]</span></p><p class="c2"><span style="overflow: hidden; display: inline-block; margin: 0.00px 0.00px; border: 0.00px solid #000000; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px); width: 1022.00px; height: 948.00px;"><img alt="" src="images/image10.png" style="width: 1022.00px; height: 948.00px; margin-left: 0.00px; margin-top: 0.00px; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px);" title=""></span></p><p class="c2 c6"><span class="c3"></span></p><h1 class="c15" id="h.p7r4f39p0lg0"><span class="c9">Summary</span></h1><p class="c2"><span>Some of the results of these calculations are effectively orthogonal to the predictors most commonly used. We have found them useful for prioritizing candidates for additional investigation that are not clearly implicated by other methods. </span><span>Summarizing the data is not trivial,</span><sup><a href="#cmnt2" id="cmnt_ref2">[b]</a></sup><sup><a href="#cmnt3" id="cmnt_ref3">[c]</a></sup><span class="c3">&nbsp;so we create a spreadsheet highlighting our conclusions.</span></p><p class="c2 c6"><span class="c3"></span></p><p class="c2 c6"><span class="c3"></span></p><p class="c2"><span class="c8">Feedback</span><span class="c3">&nbsp;- We are continuously working to improve these methods, and welcome your input regarding problems and limitations you encounter. Feel free to contact us using the information in the publication. &nbsp;</span></p><p class="c2 c6"><span class="c3"></span></p><div class="c18"><p class="c7"><a href="#cmnt_ref1" id="cmnt1">[a]</a><span class="c3">We will create a list that:</span></p><p class="c7"><span class="c3">- is NOT a real case (we must modify our example so it has ZERO PHI)</span></p><p class="c7"><span class="c3">- shows a decent digenic pair (ideally with a direct contact, maybe even co-crystallized)</span></p><p class="c7"><span class="c3">- shows a good ddG in a good structure</span></p><p class="c7"><span class="c3">- shows a PathProx with good score, ROC and PR curves, ideally in the same variant as the ddG</span></p><p class="c7"><span class="c3">- (in a perfect world, that variant would also have a borderline CADD/GERP/REVEL score)</span></p><p class="c7"><span class="c3">- also contains some *bad* models and results for comparison.</span></p></div><div class="c18"><p class="c7"><a href="#cmnt_ref2" id="cmnt2">[b]</a><span class="c3">Here we may include a sample &quot;final summary spreadsheet&quot; like the ones we present at UDN meetings.</span></p><p class="c7"><span class="c3">Long-term goal: The pipeline will produce these for us, using rules we provide to Chris. But we can&#39;t expect that for this draft.</span></p></div><div class="c18"><p class="c7"><a href="#cmnt_ref3" id="cmnt3">[c]</a><span class="c3">may --&gt; could(?)</span></p></div></body></html>