<!doctype html>
<html lang="en">
<head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <title>Tutorial</title>
    <link href="https://cdn.jsdelivr.net/npm/bootstrap@5.2.3/dist/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-rbsA2VBKQhggwzxH7pPCaAqO46MgnOM80zW1RWuH61DGLwZJEdK2Kadq2F9CUG65" crossorigin="anonymous">
    <link href="./css/vustruct.css" rel="stylesheet">
</head>
<body>
<div class="container">
    <?php echo file_get_contents("./navbar.html") ?>

<h2>Tutorial</h2>
    <div class="accordion" id="accordionExample">
        <div class="accordion-item">
            <h2 class="accordion-header">
                <button class="accordion-button" type="button" data-bs-toggle="collapse" data-bs-target="#collapseBackground" aria-expanded="false" aria-controls="collapseBackground">
                    Background
                </button>
            </h2>
            <div id="collapseBackground" class="accordion-collapse collapse" data-bs-parent="#accordionExample">
                <div class="accordion-body">
                    <p>Many tools have been developed to predict the effect of missense variants in the genome.
                        VUStruct is provided to additionally analyze factors that are often excluded by other tools, such as
                        thermodynamic destabilization of the resulting protein structure, 3D colocalization of the variant with
                        clusters of pathogenic mutations in the same protein, potential digenic effects arising from pairwise
                        interactions of variants in different genes, and other factors described below. For details about how the
                        various methods work, please visit the Calculation Details menu above.</p>
                    <p>In this tutorial, we shall begin with an example list of variants of unknown significance (VUS) as input.
                        These variants come from the sequencing results of a fictional case for diagnosis of an unknown disease.
                        This tutorial will walk you through preparation and submission of the input data,
                        and the interpretation of the results generated by the system.</p>

                     </div>
            </div>
        </div>
        <div class="accordion-item">
            <h2 class="accordion-header">
                <button class="accordion-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseInputs" aria-expanded="false" aria-controls="collapseInputs">
                    Gather/Prepare Input Files
                </button>
            </h2>
            <div id="collapseInputs" class="accordion-collapse collapse" data-bs-parent="#accordionExample">
                <div class="accordion-body">
                    <p>VUStruct is launhed from a file of candidate variants for a single case. Each case typically includes 5-50
                        variants preselected from WES or WGS of the proband. This selection removes variants of unreliable quality,
                        variants too common to be associated with the proband’s phenotype, variants associated with conditions
                        unrelated to the patient’s condition, and known benign variants. In the UDN workflow, this step is usually
                        performed by experts in clinical genetics, using standard software tools like VEP and databases of
                        biomedical literature like OMIM.</p>
                    <p>The length of the input lists for this webserver is limited to 100 variants, because the ddG calculations are
                        computationally intensive. But please contact us if you want to submit a larger set.</p>
                    <p>
                        Submit only de-identified data. You will be given a link to access the results that is on the public internet,
                        however difficult to guess.
                    </p>
                    <p>
                        Include inheritance information if you have it. Typically the proband’s parents are included in the WES and WGS
                        testing, to ascertain the inheritance of the variants, and match it with the segregation of the phenotype.
                        Sometimes additional relatives are included, e.g. siblings who are affected or unaffected.
                    </p>
                    <p>The VUStruct launch form accepts a variety of input formats, all of which are preprocessed to "vustruct.csv" format.</p>

<div class="table-responsive">
<table class="table table-hover table-bordered" style="font-size:60%">
                <caption>Sample table from a vustruct.csv file</caption>

<!-- 
	<colgroup width="44"></colgroup>
	<colgroup width="84"></colgroup>
	<colgroup width="49"></colgroup>
	<colgroup width="76"></colgroup>
	<colgroup width="55"></colgroup>
	<colgroup width="126"></colgroup>
	<colgroup width="661"></colgroup>
	<colgroup width="69"></colgroup>
	<colgroup width="116"></colgroup>
	<colgroup width="126"></colgroup>
	<colgroup width="59"></colgroup>
	<colgroup width="76"></colgroup>
	<colgroup width="92"></colgroup> -->
	<tr>
		<th height="17" align="left">index</th>
		<th align="left">gene</th>
		<th align="left">chrom</th>
		<th align="left">pos</th>
		<th align="left">change</th>
		<th align="left">effect</th>
		<th align="left">transcript</th>
		<th align="left">unp</th>
		<th align="left">refseq</th>
		<th align="left">mutation</th>
		<th align="left">genome</th>
		<th align="left">inheritance</th>
		<th align="left">zygosity</th>
	</tr>
	<tr>
		<td height="17" align="right" sdval="1" sdnum="1033;">1</td>
		<td align="left">COA5</td>
		<td align="left">chr2</td>
		<td align="right" sdval="98608380" sdnum="1033;">98608380</td>
		<td align="left">G/A</td>
		<td align="left">missense</td>
		<td align="left">ENST00000328709.8</td>
		<td align="left">Q86WW8</td>
		<td align="left">NM_001008215.2</td>
		<td align="left">P9L</td>
		<td align="left">GRCh38</td>
		<td align="left">father</td>
		<td align="left">heterozygous</td>
	</tr>
	<tr>
		<td height="17" align="right" sdval="2" sdnum="1033;">2</td>
		<td align="left">NAGA</td>
		<td align="left">chr22</td>
		<td align="right" sdval="42067209" sdnum="1033;">42067209</td>
		<td align="left">C/T</td>
		<td align="left">missense</td>
		<td align="left">ENST00000396398.8;ENST00000402937.1;...</td>
		<td align="left">P17050</td>
		<td align="left">NM_000262.2</td>
		<td align="left">D136N</td>
		<td align="left">GRCh38</td>
		<td align="left">mother</td>
		<td align="left">heterozygous</td>
	</tr>
	<tr>
		<td height="17" align="right" sdval="3" sdnum="1033;">3</td>
		<td align="left">SMYD5</td>
		<td align="left">chr2</td>
		<td align="right" sdval="73218934" sdnum="1033;">73218934</td>
		<td align="left">C/A</td>
		<td align="left">missense</td>
		<td align="left">ENST00000389501.9</td>
		<td align="left">Q6GMV2</td>
		<td align="left">NM_006062.2</td>
		<td align="left">A57E</td>
		<td align="left">GRCh38</td>
		<td align="left">father</td>
		<td align="left">heterozygous</td>
	</tr>
	<tr>
		<td height="17" align="right" sdval="4" sdnum="1033;">4</td>
		<td align="left">SMARCAL1</td>
		<td align="left">chr2</td>
		<td align="right" sdval="216447069" sdnum="1033;">216447069</td>
		<td align="left">G/A</td>
		<td align="left">missense</td>
		<td align="left">ENST00000357276.9;ENST00000358207.9;...</td>
		<td align="left">Q9NZC9</td>
		<td align="left">NM_001127207.1</td>
		<td align="left">A588T</td>
		<td align="left">GRCh38</td>
		<td align="left">mother</td>
		<td align="left">heterozygous</td>
	</tr>
	<tr>
		<td height="17" align="right" sdval="5" sdnum="1033;">5</td>
		<td align="left">ONECUT1</td>
		<td align="left">chr15</td>
		<td align="right" sdval="52789080" sdnum="1033;">52789080</td>
		<td align="left">G/C</td>
		<td align="left">missense</td>
		<td align="left">ENST00000305901.7</td>
		<td align="left">Q9UBC0</td>
		<td align="left">NM_004498.2</td>
		<td align="left">R269G</td>
		<td align="left">GRCh38</td>
		<td align="left">father</td>
		<td align="left">heterozygous</td>
	</tr>
	<tr>
		<td height="17" align="right" sdval="6" sdnum="1033;">6</td>
		<td align="left">SORD</td>
		<td align="left">chr15</td>
		<td align="right" sdval="45069018" sdnum="1033;">45069018</td>
		<td align="left">CG/C</td>
		<td align="left">Frameshift deletion</td>
		<td align="left"><br></td>
		<td align="left">Q00796-1</td>
		<td align="left">NM_003104.6</td>
		<td align="left">Frameshift deletion</td>
		<td align="left">GRCh38</td>
		<td align="left">mother</td>
		<td align="left">homozygous</td>
	</tr>
	<tr>
		<td height="17" align="right" sdval="7" sdnum="1033;">7</td>
		<td align="left">SEC61A2</td>
		<td align="left">chr10</td>
		<td align="right" sdval="12085047" sdnum="1033;">12085047</td>
		<td align="left"><br></td>
		<td align="left">Deletion</td>
		<td align="left"><br></td>
		<td align="left"><br></td>
		<td align="left"><br></td>
		<td align="left">Deletion</td>
		<td align="left">GRCh38</td>
		<td align="left">de novo</td>
		<td align="left">heterozygous</td>
	</tr>
</table>
</div>

                    <ul>
                        <li><a href="./tutorial/EXAMPLE07_vustruct.csv">EXAMPLE07_vustruct.csv</a>: Variants and Gene Names in Input variants and Gene Names (with inheritance information) in Vanderbilt UDN case spreadsheet format.</li>
                        <li>Separate VCF and Gene List files:
                            <ul>
                                <li><a href="./tutorial/SampleCase.vcf">SampleCase.vcf</a>: The Genomic Coordinates of the variants in standard Variant Call Format</li>
                                <li><a href="./tutorial/SampleCase.txt">SampleCase.txt</a>: A .txt file of Gene names and inheritance information</li>
                            </ul>
                        </li>
                    </ul>
                </div>
            </div>
        </div>
        <div class="accordion-item">
            <h2 class="accordion-header">
                <button class="accordion-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseLaunch" aria-expanded="false" aria-controls="collapseLaunch">
                    Launching the VUStruct calculation pipeline
                </button>
            </h2>
            <div id="collapseLaunch" class="accordion-collapse collapse" data-bs-parent="#accordionExample">
                <div class="accordion-body">
                    <p>
                        Go to the VUStruct launch page, which can be clicked from the menu at the top of this page.  You will see an
                        input form with a SUBMIT button at the bottom.  Enter your Case Label in the top field and then select a Data Format
                        of either 'VCF GRCh38' or 'Vanderbilt UDN Case Spreadsheet'.
                        If you are proceeding with the Vanderbilt format .xlsx input data, you will not see the Additional Digenic entry box, as gene
                        names are pulled from the spreadsheet itself.</p>

                    <img src="images/VUStructLaunchFormScreenShot.png" style="width:75%;height:auto;display:block;margin-left:auto;margin-right:auto;border: #0000ff 2px solid" />

                    <br>

                    <p>To launch, click the blue/green button the bottom, 'Submit VUStruct Case to Computer Cluster'.   Press the green Submit button at the bottom of the page.
                        You will receive a link to a page where the results will appear. You can bookmark that link.
                        The duration of the calculation depends primarily on the size of the proteins.
                        If the longest protein in your list is under 200 residues in length, the calculation can
                        finish in as little as 4-6 hours. For proteins >1000 residues in length,
                        the calculation can take more than 3 days to finish. </p>

                    </p>                </div>
            </div>
        </div>
        <div class="accordion-item">
            <h2 class="accordion-header">
                <button class="accordion-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseInterpretingResults" aria-expanded="false" aria-controls="collapseInterpretingResults">
                    Interpreting Results
                </button>
            </h2>
            <div id="collapseInterpretingResults" class="accordion-collapse collapse" data-bs-parent="#accordionExample">
                <div class="accordion-body">
                    <h5>General</h5>
<p>The results page for a calculation looks like this below. The top table shows a summary of the predictions
for all variants submitted.The DiGePred predictions are shown in a heat map linked from
“DiGePred Analysis,” and summarized in the lower table.
</p><p>YEach line in the top table shows the variant information, and a summary of the PathProx and ddG calculations.
Note that the summary does not give a full picture of the predictions, because it combines
calculations performed on multiple isoforms, and multiple structural models for each variant.
You should look at each one in detail as shown below:
</p><p>You should first expand the rows for each variant of interest by clicking on the “collapsable” icon
(looks like a greater-than sign) at the left of each row. Example highlighted in red box here
</p>
                    <img src="tutorial/images/image9.png" />
<br><p>Once that row is expanded, you will see the links in the first column go to a “Report” for the variant as it
                    appears in each curated transcript (and corresponding protein isoform). Example highlighted in
                    red below [fig 05.png]. These are important links, because they will take you to the detailed
                    prediction for the calculations. Again, you should not rely on the summary alone. Click on a
                    “Report” link to see the results page for that variant.</p>
                    <img src="tutorial/images/image8.png" />
                    <br</br>
                    <h5>Report Pages</h5>
                    <p>
                    Each Report link will take you to another page that is specific for that variant in one
                    transcript/protein isoform. The calculations can be performed multiple times, depending on how
                    many structural models are available for the protein (which include the variant position).
                    An example Report page is shown below. Helpful features to note on this Report page are:</p>

<ul><li>Summary information at the top, including links back to the main “case” page, the gene information at Medline+, and the protein information at Uniprot.</li>
<li>A graphical view of the protein sequence, with PFAM domains highlighted, and showing the location of the variant of interest. This quickly shows whether the variant falls in an annotated domain, between domains, or near the protein termini. This is a quick view; for greater detail, you can look at the Interpro site at EMBL-EBI.</li>
<li>Rate4Site score for this position in the protein (for details, see the documentation page.)</li>
<li>Cosmis score for this position in the protein (for details, see the documentation page.)</li>
<li>Structure Summary table lists all structures used for the calculations. The PathProx and ddG calculations are run for each structure. There may be experimental structures from the PDB, including structures generated by X-ray crystallography, NMR solution spectroscopy, and CryoEM techniques. There also may be models gathered from the AlphaFold2 database (the names start with “AF-”). There may be models taken from ModBase at Salilab (the names start with the RefSeqID “ENSP…” ). And models from SWISS-MODEL repo at Expasy (the names start with the UniProt ID, e.g. Q13825). Some of these structures and models will contain the full-length protein, sometimes in multimeric states, others contain only domains or a fragment of the protein. The structures and models may vary widely in quality and relevance, so you will want to consider which one is the best for your purpose.</li>
<li>The left margin contains shortcuts to the section of the page that contains the results for each structure listed in the Structure Summary table. Each section contains details of the PathProx and ddG calculations. To be clear: the links in the table take you to the source of each structural model, whereas the shortcuts in the left sidebar take you to the results of the calculations lower down on this page. Click on one of the shortcuts to go to that section. (Highlighted in red here.)</li>
</ul>

<img src="tutorial/images/image3.png" />
<br>
<h5>Example section for a structural model<h5><p>
Each section contains results for one of the structural models. The results include:
</p>
<ul>
    <li>PathProx calculations using variants from ClinVar as “Pathogenic”</li>
    <li>PathProx calculations using variants from COSMIC as “Pathogenic”</li>
    <li>a 3D display (using NGL viewer) of the variant in the structural context</li>
<li>and 14 thumbnails detailing the variant distributions and statistics for the PathProx calculations</li>
</ul>
    <p>
Here are shown the results of calculations using the model from ModBase. [You can see that in this case, the ddG_monomer calculation gave a very high score, >46 kcal/mol, in contrast with the ddG_cartesian calculation which did not give a score.We launch both protocols because sometimes one of them fails to complete, due to problems such as e.g. chainbreaks in the input structure. Below that, the PathProx score of 0.27 is shown under “ClinVar Results,” which quantifies the colocalization of the variant compared to distributions of pathogenic vs benign variants (for details and references, see CalculationDetails->PathProx ).
        <img src="tutorial/images/image4.png" />
    </p>

    <h5>Structure display</h5>
    <p>The variant is shown in the context of the structural model using NGL viewer. This provides great flexibility in
    displaying the data, even permitting you to save a figure. For example, use the menu option Variants &rarr; ClinVar
    to show the locations of the pathogenic variants. Then use Colors rarr; ClinVar PathProx to paint the ribbons
        with the corresponding scores at each residue. It becomes clear why this location scored highly,
        especially if you contrast that with the display of Variants rarr; GnomAD. You can save your figure with File &rarr; “Export Image”
        </p>

    <img src="tutorial/images/image5.png" />
    <br>
    <h5>Thumbnails of PathProx Statistics</h5>
    <p>The 3D viewer is followed by additional images showing the distributions of variants compared to the
    statistics of the null distribution used by PathProx</p>
    <img src="tutorial/images/image6.png" />

    <p>It can be useful to view the ROC curve and the PR curve for the PathProx calculations.
        The score itself can be more or less convincing, depending on the number and distribution of the pathogenic
        variants and the neutral variants. You can see in the PathProx Clinvar ROC below, that the classification
        confidence is suggestive, but not overwhelmingly strong. Does that match your intuition from looking at the
        distributions in the NGL viewer above? For a better understanding of spatial constraint, be sure to read
        (<a href="https://pubmed.ncbi.nlm.nih.gov/29455857/">Sively et al. 2018</a>)
        <img src="tutorial/images/image2.png" />

    </div>
<h5>DiGePred results</h5>
    <p>The genes you provide in your list are evaluated for the potential of digenic disease interactions.
        This is done using a predictor called <a href="https://pubmed.ncbi.nlm.nih.gov/34529933/">DiGePred</a>,
        a random forest classifier for identifying candidate digenic disease gene pairs by features derived from
        biological networks, genomics, evolutionary history, and functional annotations. Go back to the “case page”
        that summarizes all the results for this VUStruct run. You’ll see a summary table of the DiGePred scores.
        Just above that is the link to the detailed results (highlighted in red below):
<img src="tutorial/images/image1.png" />
<p>
        The digenic predictions are displayed as a heat map. Each cell represents the potential interactions
        between two of the genes in the list. The matrix is symmetric (redundant). The legend on the right
        hand side indicates the colors that the cell borders will be drawn, used to highlight the inheritance
        information (if provided with the input data). The tabs at the bottom are used to select subsets of
        the information used as support for the predicted interactions, and change the display of the heatmap.
        Moving the mouse cursor over individual cells displays a pop-up tip that summarizes the score, and
        the information used to support that score.
    </p>
<img src="tutorial/images/image10.png" />
                <br>
<h5>Summary</h5>
<p>Some of the results of these calculations are effectively orthogonal to the predictors most commonly used. We have
found them useful for prioritizing candidates for additional investigation that are not clearly implicated by other
methods. Summarizing the data is not trivial, so we create a spreadsheet highlighting our conclusions.
</p>



</div>

    <script src="https://cdn.jsdelivr.net/npm/bootstrap@5.2.3/dist/js/bootstrap.bundle.min.js" integrity="sha384-kenU1KFdBIe4zVF0s0G1M5b4hcpxyD9F7jL+jjXkk+Q2h455rYXK/7HAuoJl+0I4" crossorigin="anonymous"></script>
<script src="js/includeHTML.js">
</script>

<script>
    includeHTML();
</script>
</body>
</html>
